tessel/t2-cli
# Adds SDK installation and Rust local compilation.

Adds the following capabilities:

* `t2 sdk install` — installs T2 SDK and rustlib.
* `t2 sdk remove` — removes the above.
* `t2 run Cargo.toml` — locally compiles a Cargo crate and deploys on Tessel.

## Usage

Use this branch of `t2-cli`.

In a new directory, run `t2 init --lang=rust`.

Run `t2 sdk install`. If it errors out, make sure you're using rust stable v1.11.0 or v1.12.0. If you have `rustup` on your machine, run `rustup override set stable`.

Run `t2 run Cargo.toml`. It should output the following and start blinking:

```
$ t2 run Cargo.toml
INFO Looking for your Tessel...
INFO Connected to Tessel-02A34C9AB2FE.
    Updating registry `https://github.com/rust-lang/crates.io-index`
   Compiling libc v0.2.16
   Compiling cfg-if v0.1.0
   Compiling unix_socket v0.5.0
   Compiling tessel v0.2.0
   Compiling blinky v0.0.1 (file:///Users/tim/tcr/test/rust-blinky)
    Finished release [optimized] target(s) in 3.38 secs
INFO Writing project to RAM on Tessel-02A34C9AB2FE (338.432 kB)...
INFO Deployed.
INFO Running blinky...
I'm blinking! (Press CTRL + C to stop)
```

## Testing

Setup: run `cargo-tessel sdk uninstall` to reset your local context.

**NOTE:** The remote server *compilation* may fail until this is updated. Interpret a lot of `__stack_chk_guard` errors as `--rustcc` working properly.

* Run `t2 init --lang=rust` in a new directory to get a blinky example.
* Run `t2 run blinky` when `rustc` is not available. You should be informed A) you should install `rustc` and Cargo and B) you can run `t2 run Cargo.toml --rustcc` to use the Rust Cross Compiler. We exit with an error code.
* Run `t2 run blinky --rustcc` when `rustc` is not available. The remote compilation server should be used and succeed, and you will see the blinky code run on Tessel. (You can `mv /Users/tim/.cargo/bin/rustc` to some other path to simulate rustc not being installed. Make sure to move it back after.)
* Run `t2 run blinky` when `rustc` *is* available. You should be informed A) **the Rust SDK needs to be installed to compile locally** and B) you can run `t2 run Cargo.toml --rustcc` to use the Rust Cross Compiler. We exit with an error code.
* Run `cargo tessel sdk install` when your `rustc` version is nightly or < 1.11.0. It should install the Tessel build tools. (If you have `rustup`, run `rustup override set nightly` in your blinky dir first). It should fail to install the MIPS libstd with a warning to install 1.11.0 or greater stable. We exit with an error code.
* Run `t2 run blinky`. You should be informed A) **the MIPS libstd needs to be installed to compile locally** and B) you can run `t2 run Cargo.toml --rustcc` to use the Rust Cross Compiler. We exit with an error code.
* Run `cargo tessel sdk install` when your `rustc` version is >= 1.11.0 and is stable.  (If you have rustup, run `rustup override set stable` in your blinky dir first). It should install the Tessel build tools and the MIPS libstd successfully.
* Run `t2 run blinky`. The compilation should be successful, using the locally installed SDK and cargo to compile the blinky code, then run on Tessel.
* Run `t2 run blinky --rustcc`. The compilation should be successful, using the remote server to compile the blinky code, then run on Tessel.
